frontend/src/pages/Compare.tsx

```diff
@@ -14,56 +14,115 @@
+        if (data.length < 2 || data.length > 10) {
+          setCompareVersion(CompareVersion.InvalidRunCount);
+        } else {
+          const v2runs = data.filter(run =>
+            run.run?.pipeline_spec?.hasOwnProperty('pipeline_manifest'),
+          );
+          if (v2runs.length === 0) {
+            setCompareVersion(CompareVersion.V1);
+          } else if (v2runs.length === data.length) {
+            setCompareVersion(CompareVersion.V2);
+          } else {
+            setCompareVersion(CompareVersion.Mixed);
+          }
+        }
+      },
+    },
+  );
+
+  useEffect(() => {
+    // Update banner based on error, feature flag, run versions, and run count.
+    if (isError) {
+      (async function() {
         const errorMessage = await errorToMessage(error);
-        props.updateBanner({
+        updateBanner({
           additionalInfo: errorMessage ? errorMessage : undefined,
           message: `Error: failed loading ${runIds.length} runs. Click Details for more information.`,
           mode: 'error',
         });
-      },
-      onSuccess: () => props.updateBanner({}),
-    },
-  );
+      })();
+    } else if (
+      isFeatureEnabled(FeatureKey.V2_ALPHA) &&
+      compareVersion === CompareVersion.InvalidRunCount
+    ) {
+      updateBanner({
+        additionalInfo:
+          'At least two runs and at most ten runs must be selected to view the Run Comparison page.',
+        message:
+          'Error: failed loading the Run Comparison page. Click Details for more information.',
+        mode: 'error',
+      });
+    } else if (isFeatureEnabled(FeatureKey.V2_ALPHA) && compareVersion === CompareVersion.Mixed) {
+      updateBanner({
+        additionalInfo:
+          'The selected runs are a mix of V1 and V2.' +
+          ' Please select all V1 or all V2 runs to view the associated Run Comparison page.',
+        message:
+          'Error: failed loading the Run Comparison page. Click Details for more information.',
+        mode: 'error',
+      });
+    } else if (isFeatureEnabled(FeatureKey.V2_ALPHA) && compareVersion !== CompareVersion.V1) {
+      // Clear the banner unless the V1 page is shown, as that page handles its own banner state.
+      updateBanner({});
+    }
+  }, [compareVersion, isError, error, updateBanner, runIds.length]);
 
-  if (isError || data === undefined) {
+  if (isError) {
     return <></>;
   }
 
-  // Display the V2 Compare page if all selected runs are V2.
-  if (
-    isFeatureEnabled(FeatureKey.V2_ALPHA) &&
-    isSuccess &&
-    data &&
-    data.every(run => run.run?.pipeline_spec?.hasOwnProperty('pipeline_manifest'))
-  ) {
-    return <CompareV2 />;
-  } else {
+  if (!isFeatureEnabled(FeatureKey.V2_ALPHA)) {
     return <CompareV1 {...props} />;
+  } else {
+    if (compareVersion === CompareVersion.V2) {
+      return <CompareV2 />;
+    } else if (compareVersion === CompareVersion.V1) {
+      return <CompareV1 {...props} />;
+    } else if (
```

Comment:
NIT: Same, I guess the `else if` condition here is uncessary. You can return <></> directly if the condition checks above all return false.